cmd/nerdctl/compose.go

```diff
@@ -81,53 +82,44 @@ func newComposeCommand() *cobra.Command {
 	return composeCommand
 }
 
-func getComposer(cmd *cobra.Command, client *containerd.Client, globalOptions *types.GlobalCommandOptions) (*composer.Composer, error) {
-	nerdctlCmd, nerdctlArgs := globalFlags(cmd)
-	projectDirectory, err := cmd.Flags().GetString("project-directory")
+func getComposeOptions(cmd *cobra.Command, client *containerd.Client, globalOptions *types.GlobalCommandOptions) (*composer.Options, error) {
+	options := &composer.Options{}
+	options.NerdctlCmd, options.NerdctlArgs = globalFlags(cmd)
+	var err error
+	options.ProjectDirectory, err = cmd.Flags().GetString("project-directory")
 	if err != nil {
 		return nil, err
 	}
-	envFile, err := cmd.Flags().GetString("env-file")
+	options.EnvFile, err = cmd.Flags().GetString("env-file")
 	if err != nil {
 		return nil, err
 	}
-	projectName, err := cmd.Flags().GetString("project-name")
+	options.Project, err = cmd.Flags().GetString("project-name")
 	if err != nil {
 		return nil, err
 	}
-	debugFull := globalOptions.DebugFull
-	snapshotter := globalOptions.Snapshotter
-	files, err := cmd.Flags().GetStringArray("file")
+	options.DebugPrintFull, err = cmd.Flags().GetBool("debug-full")
 	if err != nil {
 		return nil, err
 	}
-	insecure := globalOptions.InsecureRegistry
-	cniPath := globalOptions.CNIPath
-	cniNetconfpath := globalOptions.CNINetConfPath
-	hostsDirs := globalOptions.HostsDir
-	experimental := globalOptions.Experimental
-
-	o := composer.Options{
-		Project:          projectName,
-		ProjectDirectory: projectDirectory,
-		ConfigPaths:      files,
-		EnvFile:          envFile,
-		NerdctlCmd:       nerdctlCmd,
-		NerdctlArgs:      nerdctlArgs,
-		DebugPrintFull:   debugFull,
-		Experimental:     experimental,
+	options.ConfigPaths, err = cmd.Flags().GetStringArray("file")
+	if err != nil {
+		return nil, err
+	}
+	options.Experimental, err = cmd.Flags().GetBool("experimental")
+	if err != nil {
+		return nil, err
 	}
```

Comment:
Sorry for the back and forth, but I think the original format of the first half function should serve us well right? (e.g., read cmd flags, and create `composer.Options` altogether) (so suggest not change the 1st half)

Meanwhile, as you can see in this whole function, it contains two types of code: 

1. cobra related flag process: it's not related to compose implementation details so we should keep it in `cmd` side; 
2. compose implementation details: e.g., those `o.NetworkExists = ..., o.VolumeExists = ...` etc. they are not related to cobra/cmd after we get the option _values_ from the 1st part.

What I meant in my comment (https://github.com/containerd/nerdctl/pull/1791#issuecomment-1370041137) is that we should spit this two parts into two functions (so cobra logic stays in `cmd` and implementation goes to `pkg`): 

1. `getComposeOptions(cmd *cobra.Command) composer.Options`: basically [the 1st half](https://github.com/containerd/nerdctl/blob/c3e6ceb0de52e681f1217d6024af97bedece801d/cmd/nerdctl/compose.go#L85-L119). after this call, `composer.Options` contains all option values needed for the implementation.
2. `GetComposer(opt composer.Options, client *containerd.Client, stdout, stderr io.Writer) (*composer.Composer, error)`: basically [the 2nd half](https://github.com/containerd/nerdctl/blob/c3e6ceb0de52e681f1217d6024af97bedece801d/cmd/nerdctl/compose.go#L121-L215). It first adds (our default) implementation to `option` and then returns a `composer`. If we want to enable callers to provide other implementations via `composeoption`, we should check if the field if nil.

```golang
if o.EnsureImage == nil {
    o.EnsureImage = func(....
}
```

After this change, the workflow for a compose command to obtain a `composer` will be:

```golang
    // get option values from cmd
	options, err := getComposeOptions(cmd, client, globalOptions)
	if err != nil {
		return err
	}
    // get composer
	c, err := composer.GetComposer(options, client, cmd.OutOrStdout(), cmd.ErrOrStdErr())
``` 

Let me know if you have other questions. @Laitr0n 

@AkihiroSuda, @Zheaoli WDYT? (does this make sense :))